src/perf_map_poller.rs

```diff
@@ -0,0 +1,152 @@
+use std::{
+    collections::HashMap,
+    fmt::{self, Formatter},
+    sync::{Arc, Condvar, Mutex},
+    thread,
+    time::Duration,
+};
+
+use crossbeam_channel::{Sender, TrySendError};
+use mio::{unix::SourceFd, Events, Interest, Poll, Token};
+use nix::errno::Errno;
+use slog::crit;
+
+use crate::{
+    maps::{PerCpu, PerfEvent, PerfMap},
+    PerfChannelMessage, LOGGER,
+};
+
+pub struct PerfMapPoller {
+    poll: Poll,
+    tokens: HashMap<Token, PerfMap>,
+}
+
+impl PerfMapPoller {
+    pub fn new(
+        perfmaps: impl Iterator<Item = PerfMap>,
+        polling_signal: Arc<(Mutex<bool>, Condvar)>,
+    ) -> Result<Self, InitError> {
+        let poll = Poll::new().map_err(InitError::Creation)?;
+        let registry = poll.registry();
+
+        let tokens = perfmaps
+            .map(|p| {
+                let token = Token(p.ev_fd as usize);
+                registry
+                    .register(&mut SourceFd(&p.ev_fd), token, Interest::READABLE)
+                    .map(|_| (token, p))
+            })
+            .collect::<Result<_, _>>()
+            .map_err(InitError::Registration)?;
+
+        {
+            // now that the perfmap fd's are registered, we can signal to the main thread that
+            // event polling is ready.
+            let (lock, cvar) = &*polling_signal;
```

Comment:
I'd like @rc-dbogle to put his eyes on the fds here and make sure we're maintaining the contract properly. We've had issues with that before. It looks fine to me but he has more experience with this.

Reply:
Are you referring to the contract about when to set the mutex/cvar or registering the fds with the poller?

Reply:
I'm referring to making sure fds are being copied and duplicated across threads appropriately, which was the cause of that cleanup bug a while back.